core/domain/prod_validation_jobs_one_off.py

```diff
@@ -4942,6 +4942,128 @@ def _get_custom_validation_functions(cls):
+                % (item.id,))
+
+    @classmethod
+    def _validate_status(cls, item):
+        """Validates the fields of the item relating to the status field.
+
+        Args:
+            item: improvements_models.TaskEntryModel.
+        """
+        if item.status == improvements_models.STATUS_OPEN:
+            if item.closed_by:
+                cls.errors['status check'].append(
+                    'Task id %s: Status is open but closed_by is "%s", '
+                    'should be empty.' % (item.id, item.closed_by))
+            if item.closed_on:
+                cls.errors['status check'].append(
+                    'Task id %s: Status is open but closed_on is "%s", '
+                    'should be empty.' % (item.id, item.closed_on))
+        elif item.status == improvements_models.STATUS_RESOLVED:
+            if item.closed_by is None:
+                cls.errors['status check'].append(
+                    'Task id %s: Status is resolved but closed_by is not '
+                    'set' % (item.id,))
+            if item.closed_on is None:
+                cls.errors['status check'].append(
+                    'Task id %s: Status is resolved but closed_on is not '
+                    'set' % (item.id,))
+
+    @classmethod
+    def _validate_task_type(cls, item):
+        """Validates the fields of the item relating to the task_type field.
+
+        Args:
+            item: improvements_models.TaskEntryModel.
+        """
+        if item.task_type in (
+                improvements_models.TASK_TYPE_HIGH_BOUNCE_RATE,
+                improvements_models.TASK_TYPE_NEEDS_GUIDING_RESPONSES,
+                improvements_models.TASK_TYPE_SUCCESSIVE_INCORRECT_ANSWERS,
+                improvements_models.TASK_TYPE_INEFFECTIVE_FEEDBACK_LOOP):
+            if item.entity_type != improvements_models.ENTITY_TYPE_EXPLORATION:
+                cls.errors['task type check'].append(
+                    'Task id %s: Task type %s is not supported for %s '
+                    'entities' % (item.id, item.task_type, item.entity_type))
+            if item.target_type != improvements_models.TARGET_TYPE_STATE:
+                cls.errors['task type check'].append(
+                    'Task id %s: Task type %s is not supported for %s '
+                    'targets' % (item.id, item.task_type, item.target_type))
+            if not item.target_id:
+                cls.errors['task type check'].append(
+                    'Task id %s: Task type %s needs a target but no target '
+                    'id is set' % (item.id, item.task_type))
+
+    @classmethod
+    def _validate_exploration_state_targets(cls, item):
+        """Validate that the given item contains an existing exploration state
+        name.
+
+        Args:
+            item: improvements_models.TaskEntryModel.
+        """
+        if item.entity_type != improvements_models.ENTITY_TYPE_EXPLORATION:
+            return
```

Comment:
I chose not to because it isn't an error for a task to target a non-exploration entity. The validator is forward-looking, so new types can be introduced without invalidating these checks.

Additionally, there's no need to check that "only exploration entities exist in storage" because of the `choices` argument used in `ndb.StringProperty`. It asserts that only the provided "choices" can be `put` into storage (in this PR, `ENTITY_TYPE_EXPLORATION` and `TEST_ONLY_ENTITY_TYPE` are the choices).

Ditto for `target_type`.

Reply:
No, this validator shouldn't be forward looking, in terms of what's actually checked. New types can be introduced here at the point when we add such new types to our storage layer.

I'm fine with deferring the validation to "choices" though.